### ISSUE TITLE
Profiler raises 'wrong number of arguments for constructor (ArgumentError)' for the march_hare gem

### ISSUE THREAD

**Issue by kevinwatson** [jruby/jruby#8148]
**Environment Information**

```
jruby -v
jruby 9.4.6.0 (3.1.4) 2024-02-20 576fab2c51 OpenJDK 64-Bit Server VM 17.0.10+7 on 17.0.10+7 +jit [aarch64-linux]
```

```
uname -a
Linux a2fbb69ea12d 5.15.82-0-virt #1-Alpine SMP Mon, 12 Dec 2022 09:15:17 +0000 aarch64 aarch64 aarch64 GNU/Linux
```

gem `march_hare`
env vars `JRUBY_OPTS=--profile.api`

**Expected Behavior**

The march_hare gem should be able to connect to rabbitmq and initialize its exception handler

**Actual Behavior**

```ruby
irb(main):002:0> MarchHare.connect(host: 'rabbit')
/usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:692:in `init_default_exception_handler': wrong number of arguments for constructor (ArgumentError)
	from /usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:153:in `initialize'
	from org/jruby/RubyClass.java:935:in `new'
	from /usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:132:in `connect'
	from /usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare.rb:24:in `connect'
	from (irb):2:in `evaluate'
	from org/jruby/RubyKernel.java:1265:in `eval'
	from org/jruby/RubyKernel.java:1707:in `loop'
	from org/jruby/RubyKernel.java:1414:in `catch'
	from org/jruby/RubyKernel.java:1414:in `catch'
	from /opt/jruby/lib/ruby/gems/shared/gems/irb-1.4.2/exe/irb:11:in `<main>'
	from org/jruby/RubyKernel.java:1211:in `load'
	from /opt/jruby/bin/irb:25:in `<main>'
```

https://github.com/ruby-amqp/march_hare/blob/main/lib/march_hare/exception_handler.rb#L2

This docker compose file can be used to test this behavior

```yml
# compose.yml
services:
  console:
    image: jruby:9.4-jdk17
    depends_on:
    - rabbit
  rabbit:
    image: rabbitmq:latest
    ports:
    - 5672:5672
```

Works successfully when `--profile.api` is not set

```bash
docker-compose run console bash

gem install march_hare
irb

require 'march_hare'
MarchHare.connect(host: 'rabbit')
=>
#<MarchHare::Session:0xc1f0c7b
 @addresses=["rabbit"],
 @automatic_recovery_hook=
  #<MarchHare::ShutdownListener:0x3a7d914c
   @block=#<Proc:0x5100c143 /usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:308>,
   @entity=#<MarchHare::Session:0xc1f0c7b ...>>,
 @automatically_recover=true,
...
```

Error when `--profile.api` is set

```bash
docker-compose run console bash

gem install march_hare
export JRUBY_OPTS=--profile.api
irb

require 'march_hare'
MarchHare.connect(host: 'rabbit')
/usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:692:in `init_default_exception_handler': wrong number of arguments for constructor (ArgumentError)
	from /usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:153:in `initialize'
	from org/jruby/RubyClass.java:935:in `new'
	from /usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:132:in `connect'
	from /usr/local/bundle/gems/march_hare-4.6.0-java/lib/march_hare.rb:24:in `connect'
	from (irb):2:in `evaluate'
	from org/jruby/RubyKernel.java:1265:in `eval'
	from org/jruby/RubyKernel.java:1707:in `loop'
	from org/jruby/RubyKernel.java:1414:in `catch'
	from org/jruby/RubyKernel.java:1414:in `catch'
	from /opt/jruby/lib/ruby/gems/shared/gems/irb-1.4.2/exe/irb:11:in `<main>'
	from org/jruby/RubyKernel.java:1211:in `load'
	from /opt/jruby/bin/irb:25:in `<main>'
```

**Comment by headius** [comment]
Thank you for the complete information! I suspect this is a bug in the profiler's method wrappers causing it to call along a path with the wrong number of arguments.

**Comment by headius** [comment]
Reproduced on JRuby master. It doesn't need to be able to establish a connection; it fails during the error while failing to establish a connection.

```
 $ jruby --profile.api -Xbacktrace.style=full -e "gem 'march_hare'; require 'march_hare'; MarchHare.connect(host: 'rabbit')"
Profiling enabled; ^C shutdown will now dump profile info
ArgumentError: wrong number of arguments for constructor
                   getStackTrace at java/lang/Thread.java:2417
                getBacktraceData at org/jruby/runtime/backtrace/TraceType.java:269
                    getBacktrace at org/jruby/runtime/backtrace/TraceType.java:54
                captureBacktrace at org/jruby/RubyException.java:434
                        preRaise at org/jruby/exceptions/RaiseException.java:217
                          <init> at org/jruby/exceptions/RaiseException.java:65
                          <init> at org/jruby/exceptions/Exception.java:39
                          <init> at org/jruby/exceptions/StandardError.java:38
                          <init> at org/jruby/exceptions/ArgumentError.java:38
              constructThrowable at org/jruby/RubyArgumentError.java:58
                     toThrowable at org/jruby/RubyException.java:396
                            from at org/jruby/exceptions/RaiseException.java:94
               newRaiseException at org/jruby/Ruby.java:4303
                newArgumentError at org/jruby/Ruby.java:3580
           matchConstructorIndex at org/jruby/javasupport/Java.java:780
                        forTypes at org/jruby/javasupport/Java.java:644
                          <init> at org/jruby/java/proxies/ConcreteJavaProxy.java:356
                          <init> at org/jruby/java/proxies/ConcreteJavaProxy.java:336
                splitInitialized at org/jruby/java/proxies/ConcreteJavaProxy.java:391
                          <init> at rubyobj/MarchHare//Users/headius/work/jruby/lib/ruby/gems/shared/gems/march_hare-4.6.0-java/lib/march_hare/exception_handler.rb:-1
                          <init> at rubyobj/MarchHare//Users/headius/work/jruby/lib/ruby/gems/shared/gems/march_hare-4.6.0-java/lib/march_hare/exception_handler.rb:-1
                     newInstance at jdk/internal/reflect/DirectConstructorHandleAccessor.java:62
           newInstanceWithCaller at java/lang/reflect/Constructor.java:502
                     newInstance at java/lang/reflect/Constructor.java:486
                            call at org/jruby/java/proxies/ConcreteJavaProxy.java:295
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:224
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:220
                            call at org/jruby/java/proxies/ConcreteJavaProxy.java:188
                            call at org/jruby/internal/runtime/methods/ProfilingDynamicMethod.java:65
```

**Comment by headius** [comment]
Looks like this is interacting with our Java subclass constructor logic. I thought the following patch might fix it, but it does not. I still suspect that some of these method type checks are getting confused by the intermediate `ProfilingDynamicMethod` but this one fix was not sufficient to correct the problem.

```
diff --git a/core/src/main/java/org/jruby/java/proxies/ConcreteJavaProxy.java b/core/src/main/java/org/jruby/java/proxies/ConcreteJavaProxy.java
index 759b5e1d06..a68af8a69d 100644
--- a/core/src/main/java/org/jruby/java/proxies/ConcreteJavaProxy.java
+++ b/core/src/main/java/org/jruby/java/proxies/ConcreteJavaProxy.java
@@ -37,6 +37,7 @@ import org.jruby.RubyClass;
 import org.jruby.RubyModule;
 import org.jruby.internal.runtime.AbstractIRMethod;
 import org.jruby.internal.runtime.SplitSuperState;
+import org.jruby.internal.runtime.methods.DelegatingDynamicMethod;
 import org.jruby.internal.runtime.methods.DynamicMethod;
 import org.jruby.internal.runtime.methods.JavaMethod.JavaMethodNBlock;
 import org.jruby.ir.JIT;
@@ -375,6 +376,11 @@ public class ConcreteJavaProxy extends JavaProxy {
         final CacheEntry methodEntry = base.searchWithCache(name);
         final boolean isLateral = isClassOrIncludedPrependedModule(methodEntry.sourceModule, base);
         DynamicMethod method = methodEntry.method;
+
+        // make sure to unwrap delegated methods for purposes of reinvoking
+        if (method instanceof DelegatingDynamicMethod) method = method.getRealMethod();
+
+        // detect static JCreate and use its wrapped init
         if (method instanceof StaticJCreateMethod) method = ((StaticJCreateMethod) method).oldInit;
 
         // jcreate is for nested ruby classes from a java class
```

cc @byteit101 @enebo

**Comment by headius** [comment]
I found where to fix this and have a patch coming... sorry for the delay! It was as I suspected: we did not unwrap certain methods used by concrete class construction, and later try to type check that they are expected types of methods.

**Comment by headius** [comment]
With the fix in #8211 I can run your example (until it fails to find a server, as expected):

```
$ jruby --profile.api -e "require 'march_hare'; MarchHare.connect(host: 'rabbit')" 
Profiling enabled; ^C shutdown will now dump profile info
MarchHare::ConnectionRefused: Connection refused: target host unknown (cannot be resolved). Target host list: rabbit, target virtual host: /, username: guest
  converting_rjc_exceptions_to_ruby at /Users/headius/work/jruby/lib/ruby/gems/shared/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:580
                new_connection_impl at /Users/headius/work/jruby/lib/ruby/gems/shared/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:620
               build_new_connection at /Users/headius/work/jruby/lib/ruby/gems/shared/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:612
                         initialize at /Users/headius/work/jruby/lib/ruby/gems/shared/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:170
                                new at org/jruby/RubyClass.java:935
                            connect at /Users/headius/work/jruby/lib/ruby/gems/shared/gems/march_hare-4.6.0-java/lib/march_hare/session.rb:132
                            connect at /Users/headius/work/jruby/lib/ruby/gems/shared/gems/march_hare-4.6.0-java/lib/march_hare.rb:24
                             <main> at -e:1
```

**Comment by kevinwatson** [comment]
Thanks! Much appreciated.
